### PR TITLE
Create login page

### LINKED ISSUE #31: Create login page
Create login page
- if you are not logged in, you will be taken to the login page
- else redirected to maps site to prevent tedious access

### PR DESCRIPTION
**Pull request by gannerr**
We create a login page so as to protect data and provide a layer of security against bots that may contribute falsified information to our crowdsourcing suggestion feature that allows users to correct any wrong information that we may scrape and present.
- Code changes:
   - Moved maps page code from index.html to a new file main.html, and moved login page code to index.html. It was very inconvenient that the maps page was the first page we loaded into, because there the redirect code did not let us switch from the maps page to the login page immediately (this occurs when the user is not logged in and accesses the site).
- Safety considerations: You must be logged in to see the maps page and make searches and suggest information
- UI considerations: we opted for a super minimal look with just the service name and a login button because we want people to get right to using our service!
- Login behavior: If the user is logged in already, we automatically redirect to map page. Else, go to the login page.
- Optimizations: We initially started with a servlet that returned the literal html for the login page. This approach was limited because it was difficult to ascertain login status (we had to check for the length of the html returned! smh) and expand for additional fields should we choose to store more than the user's email. As a result, we opted to create a class "LoginStatus" that contained a user's login status (that we can now quickly call and check in other functions!) and other fields like user email and the link the user would be redirected to upon logging in / logging out / making crowdsourcing suggestion feature. So much easier and expandable!
- Fun animation features: Button has pulsing (think heartbeat) animation upon hover. Title text also has small "popping" animation upon page start.
### Quick Snapshot
Screenshot 2020-08-04 at 5 48 31 PM

[TODO 1](https://github.com/googleinterns/step196-2020/issues/53): I know what you're thinking. This page is pretty ugly. Don't worry, I promise it'll look better.
[TODO 2](https://github.com/googleinterns/step196-2020/issues/34): With every login buttons comes a logout button. Let's create that logout button!

[Refer to design doc for detailed design considerations](https://docs.google.com/document/d/1nenvnmAF-OcNMX56U0LGvLItrGOa-XjLfM5LtW8Vi8Y/edit?disco=AAAAIuh_41M)

### CODE REVIEW

**Comment by gannerr** [comment]
Screenshot 2020-08-04 at 5 48 31 PM
This is the login screen. You are only here if you are not logged in. Click the "let's go" button to login.
Screenshot 2020-08-04 at 5 48 41 PM
Google login
Screenshot 2020-08-04 at 5 48 48 PM
Map page, redirected here once you logged in. Automatically redirected here if already logged in and accessing site.

**Comment by casanted** [comment]
> Screenshot 2020-08-04 at 5 48 31 PM
> This is the login screen. You are only here if you are not logged in. Click the "let's go" button to login.
> Screenshot 2020-08-04 at 5 48 41 PM
> Google login
> Screenshot 2020-08-04 at 5 48 48 PM
> Map page, redirected here once you logged in. Automatically redirected here if already logged in and accessing site.

Nice Inclusion of the screenshot!

File: capstone/src/main/webapp/loginAuth.js [hunk 1]
```diff
@@ -0,0 +1,12 @@
+function fetchLoginStatus() {
+  fetch('/login').then(response => response.text()).then((loginStatus) => {
+    //Because loginStatus will return either the logout / login message
+    //we can just check the length of return message to ascertain if
+    //we are logged in or out
+    if(loginStatus.length < 100) {
```
Comment (casanted) [inline capstone/src/main/webapp/loginAuth.js:6]: Could you please explain why the length of the login message will be < 100 if the user is logged in? I think I missed that point.

**Comment by terencedignon** [comment]
Hi Andrew, could you tag this to an issue, and add some more description to the PR? For example, 'Create a log in page to allow users to x, y, and z. Addresses the abc section of design doc (with link). 

For workflow purposes I'm going to change the assignee back to you, and then you can assign back to Carolyn and I as the comments are addressed.

Thanks!

File: capstone/src/main/java/com/google/sps/servlets/LoginServlet.java [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+      String logoutUrl = userService.createLogoutURL(urlToRedirectToAfterUserLogsOut);
+
+      response.getWriter().println("<p>Hello " + userEmail + "!</p>");
+      response.getWriter().println("<p>Logout <a href=\"" + logoutUrl + "\">here</a>.</p>");
+    } else {
+      String urlToRedirectToAfterUserLogsIn = "/index.html";
+      String loginUrl = userService.createLoginURL(urlToRedirectToAfterUserLogsIn);
+
+      response.getWriter().println("<p>Hello stranger.</p>");
+      response.getWriter().println("<p>Login <a href=\"" + loginUrl + "\">here</a>.</p>");
```
Comment (graceehuu) [inline capstone/src/main/java/com/google/sps/servlets/LoginServlet.java:44]: What is the purpose of returning this information?

File: capstone/src/main/webapp/loginAuth.js (hunk 1 above)
Comment (graceehuu) [reply capstone/src/main/webapp/loginAuth.js:6]: I agree with Carolyn, this logic could be a bit unclear. Maybe have the servlet return 'true' or 'false' and check if the response is 'true' or 'false' or something a bit more deterministic and clear

File: capstone/src/main/java/com/google/sps/servlets/LoginServlet.java (hunk 2 above)
Comment (gannerr) [reply capstone/src/main/java/com/google/sps/servlets/LoginServlet.java:44]: the user sees this printed stuff so they know what to do (click the link to login/logout). Also serves as information for fetch function to see if user is logged in or not.

File: capstone/src/main/webapp/loginAuth.js (hunk 1 above)
Comment (gannerr) [reply capstone/src/main/webapp/loginAuth.js:6]: in the login servelet there are a bunch of response.getWriter().println statements. These statements are what is ultimately stored in loginStatus. The total length of these statements for a logged out user is <100, thus we check this length to determine user login/logout status. Pretty convoluted and can definitely be optimized... lol

Comment (gannerr) [reply capstone/src/main/webapp/loginAuth.js]: updated to just using loginStatus class which contains login state for better readability

File: capstone/src/main/java/com/google/sps/servlets/LoginServlet.java (hunk 2 above)
Comment (gannerr) [reply capstone/src/main/java/com/google/sps/servlets/LoginServlet.java]: this was actually so extra so i deleted

File: capstone/src/main/webapp/loginAuth.js [hunk 3]
```diff
@@ -1,12 +1,23 @@
 function fetchLoginStatus() {
-  fetch('/login').then(response => response.text()).then((loginStatus) => {
-    //Because loginStatus will return either the logout / login message
-    //we can just check the length of return message to ascertain if
-    //we are logged in or out
-    if(loginStatus.length < 100) {
-      window.location = "/login.html";
+  fetch("/login").then(response => response.json()).then((loginStatus) => {
+    const redirectUrl = document.createElement('a');
+    redirectUrl.setAttribute('href', loginStatus.url);
```
Comment (casanted) [inline capstone/src/main/webapp/loginAuth.js:4]: Thank you for making this change! This makes it easier for us to see what is going on in your logic here. I like this method a lot!

**Review by casanted — APPROVED** [review]
LGTM!

File: capstone/src/main/webapp/loginAuth.js [hunk 4]
```diff
@@ -0,0 +1,23 @@
+function fetchLoginStatus() {
+  fetch("/login").then(response => response.json()).then((loginStatus) => {
+    const redirectUrl = document.createElement('a');
+    redirectUrl.setAttribute('href', loginStatus.url);
+    if (loginStatus.status) {
+      // if logged in, go to maps page
+      window.location = "/main.html";
+      // TODO: create logout button on page
```
Comment (terencedignon) [inline capstone/src/main/webapp/loginAuth.js:8]: Could we link this to an issue?

**Review by terencedignon — APPROVED** [review]
Looks good, thanks for adding the description. LGTMing to unblock you but please resolve the linter errors before submitting. Thanks!

File: capstone/src/main/webapp/loginAuth.js (hunk 4 above)
Comment (gannerr) [reply capstone/src/main/webapp/loginAuth.js:8]: yep

File: capstone/src/main/webapp/loginAuth.js [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+ * checks if user is logged in or not
+ */
+function fetchLoginStatus() {
+  fetch('/login').then((response) => response.json()).then((loginStatus) => {
+    const redirectUrl = document.createElement('a');
+    redirectUrl.setAttribute('href', loginStatus.url);
+    if (loginStatus.status) {
+      // if logged in, go to maps page
+      window.location = '/main.html';
+      // TODO: create logout button on page
```
Comment (gannerr) [inline capstone/src/main/webapp/loginAuth.js:11]: https://github.com/googleinterns/step196-2020/issues/34

**Pull request merged**